exonum-java-binding/testkit/src/main/java/com/exonum/binding/testkit/TestKit.java

```diff
@@ -404,21 +420,29 @@ public static Builder builder() {
 
     private EmulatedNodeType nodeType = EmulatedNodeType.VALIDATOR;
     private short validatorCount = 1;
-    private final List<Class<? extends ServiceModule>> services = new ArrayList<>();
-    private TimeProvider timeProvider;
+    private Multimap<ServiceArtifactId, ServiceSpec> services = ArrayListMultimap.create();
+    private HashMap<ServiceArtifactId, String> serviceArtifactFilenames = new HashMap<>();
+    private Path artifactsDirectory;
+    private List<TimeServiceSpec> timeServiceSpecs = new ArrayList<>();
+    private Integer serverPort;
 
-    private Builder() {
-    }
+    private Builder() {}
 
     /**
      * Returns a copy of this TestKit builder.
```

Comment:
Returns a shallow copy of this TestKit builder. (in one sentence) 

However, I am concerned that just having this documentation is not reliable. The client code must not invoke any of the numerous new methods modifying the internal data structures (timeServiceSpecs, services,  serviceArtifactFilenames).

The reason it does not already perform deep copy is TimeProvider which is not copyable. The rest of the arguments — are (or are immutable). To reduce the likelihood of errors, I'd consider:
* copying the collections (new ArrayList<>(timeServiceSpecs), etc)
* renaming the method to shallowCopy
* documenting that the remaining mutable state are time providers.